modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -295,139 +319,252 @@ public void finishFull(HybridTimestampTracker timestampTracker, UUID txId, boole
+                        (unused, throwable) -> {
+                            boolean verifiedCommit = throwable == null && commit;
+
+                            Collection<ReplicationGroupId> replicationGroupIds = new HashSet<>(enlistedGroups.keySet());
+
+                            return makeDurableFinishRequest(
+                                    observableTimestampTracker,
+                                    commitPartition,
+                                    verifiedCommit,
+                                    replicationGroupIds,
+                                    txId,
+                                    commitTimestamp,
+                                    txFinishFuture);
+                        })
+                .thenCompose(Function.identity())
+                // verification future is added in order to share proper exception with the client
+                .thenCompose(r -> verificationFuture);
+    }
 
-                // Wait for commit acks first, then proceed with the finish request.
-                tuple.waitRepFut.thenCompose(clo).handle((ignored, err) -> {
-                    if (err == null) {
-                        tuple.finishFut.complete(null);
-                    } else {
-                        tuple.finishFut.completeExceptionally(err);
+    /**
+     * Durable finish request.
+     */
+    private CompletableFuture<Void> makeDurableFinishRequest(
+            HybridTimestampTracker observableTimestampTracker,
+            TablePartitionId commitPartition,
+            boolean commit,
+            Collection<ReplicationGroupId> replicationGroupIds,
+            UUID txId,
+            HybridTimestamp commitTimestamp,
+            CompletableFuture<TransactionMeta> txFinishFuture
+    ) {
+        return inBusyLockAsync(busyLock, () -> findPrimaryReplica(commitPartition, clock.now())
+                .thenCompose(meta ->
+                        finishOnPrimary(
+                                observableTimestampTracker,
+                                commitPartition,
+                                meta.getLeaseholder(),
+                                meta.getStartTime().longValue(),
+                                commit,
+                                replicationGroupIds,
+                                txId,
+                                commitTimestamp,
+                                txFinishFuture
+                        ))
+                .handle((res, ex) -> {
+                    if (ex != null) {
+                        Throwable cause = ExceptionUtils.unwrapCause(ex);
+
+                        if (cause instanceof TransactionException) {
+                            TransactionException transactionException = (TransactionException) cause;
+
+                            if (transactionException.code() == TX_WAS_ABORTED_ERR) {
+                                updateTxMeta(txId, old -> new TxStateMeta(ABORTED, old.txCoordinatorId(), null));
```

Comment:
There should be a completion of future on `TxStateMetaFinishing`, you pass it here as `txFinishFuture`. I think it requires some dedicated method as `finalStateTransition(txId, commit)` as it happens in multiple places now.